Parse nested object

ysugimoto opened the pull request:
## Why

Now Treecreeper's markdown parser supports rich relationship property definition, so we should support to define in the markdown i.e runbook.md

## What

This PR improves markdown parser with supporting multiple relationship definition like:

```
## younger siblings

- ChildType
  propNameOne: propValueOne
  propNameTwo: propNameTwo
  ...
```

Note that this PR also includes breaking change which nested definition -- mainly relationship will return an object, not a string.

### before

`hasMany: false` case:

```
## favourite child

some child

//->  favouriteChild: 'some child'
```

`hasMany: true` case:

```
## children

* some child

//->  children: ['some child']
```

### after

`hasMany: false` case:

```
## favourite child

some child

//->  favouriteChild: { code: 'some child' }
```

`hasMany: true` case:

```
## children

* some child

//->  children: [{ code: 'some child' }]
```

## nested property validation

I defined a spec that nested property name should be a lower case, so

- some string: value -> NG
- someString: value -> OK

Of course, this rule can be looser, which accepts white space character, etc...

Trello: https://trello.com/c/KkCSNp47/158-implementing-writing-properties-on-relationships-in-runbookmd

ysugimoto commented on packages/tc-markdown-parser/__tests__/property-value.spec.js:
```diff
@@ -334,3 +348,199 @@ describe('dealing with html comments', () => {
+			expect(data.name).toEqual('name');
+			const [{ message }] = errors;
+			expect(message).toMatch(/should be lower camel case/);
+			expect(message).toMatch(/line 6/);
```
It was hard to overwrite `line` property in unified AST, so I intended to include the actual line number inside the error message

wheresrhys commented:
@chee Woudl be great if you coudl review this - we are enhancing runbook.md to
- turn into a slightly more generic parser of md into biz ops (done in a previous PR)
- add the ability to nest extra key value pairs within lists

wheresrhys commented on packages/tc-markdown-parser/lib/property-coercers/index.js:
```diff
@@ -72,6 +73,67 @@ module.exports = {
+				};
+			} catch (error) {
+				return {
+					valud: false,
```
```suggestion
					valid: false,
```

wheresrhys commented on packages/tc-markdown-parser/lib/parse-multiline-definition.js:
```diff
@@ -0,0 +1,176 @@
+   propDef: '  propNameOne: propValueOne\n',
+ }
+*/
+const dividePropertyDefinition = subdocument => {
```
Are there tools built into mdast that can do this without us having to manually slice it up @chee ?

chee replied:
thanks for inviting me to this pull request, this is all such nice code. and so wonderfully commented!!

i don't think there is something like this, i don't think `remark` even has a plugin for kramdown-style definition lists.

wheresrhys commented on packages/tc-markdown-parser/lib/parse-multiline-definition.js:
```diff
@@ -0,0 +1,176 @@
+					start: { line },
+				},
+			} = node;
+			const { code, propDef } = dividePropertyDefinition(value);
```
propDefs would be clearer?

chee commented on packages/tc-markdown-parser/lib/tree-mutators/coerce-treecreeper-properties-to-type.js:
```diff
@@ -36,8 +36,10 @@ function getCoercer({ isNested, primitiveType, propertyType }) {
+	if (isString) {
+		return isNested
+			? propertyCoercers.NestedString
+			: propertyCoercers.String;
```
Nice ^_^

ysugimoto commented on packages/tc-markdown-parser/__tests__/property-value.spec.js:
```diff
@@ -334,3 +348,199 @@ describe('dealing with html comments', () => {
+				## favourite child
+
+				example-code
+					some string: i like it
```
Yes, this is test for the error case. And now we're checking property name with regular expression at here https://github.com/Financial-Times/treecreeper/pull/103/files/87789563c0cd83e3e437601b42a467aaafc56f58#diff-ab7fbc57659cb853a1fbf6f572881d6fR74-R78

ysugimoto commented on packages/tc-markdown-parser/__tests__/property-value.spec.js:
```diff
@@ -334,3 +348,199 @@ describe('dealing with html comments', () => {
+			expect(message).toMatch(/line 6/);
+		});
+
+		it('property value must not be empty', async () => {
```
Right, but I forgot to consider about other data types. I'll add that

JakeChampion commented on packages/tc-markdown-parser/lib/parse-multiline-definition.js:
```diff
@@ -0,0 +1,176 @@
+class SyntaxError extends Error {}
+
+/*
+ This function divides from mutipile definition to code and property string e.g,
```
```suggestion
 This function divides from multiple definition to code and property string e.g,
```

JakeChampion commented on packages/tc-markdown-parser/lib/parse-multiline-definition.js:
```diff
@@ -0,0 +1,176 @@
+const CARRIAGE_RETURN = '\r';
+
+/*
+ This is tiny lexer/parser fuction for analyze multiple property definition.
```
```suggestion
 This is tiny lexer/parser function for analyze multiple property definition.
```

JakeChampion commented on packages/tc-markdown-parser/lib/parse-multiline-definition.js:
```diff
@@ -0,0 +1,176 @@
+	let propName = '';
+	let column = 0;
+
+	for (let i = 0; i < strlen; i++) {
```
This could be replaced with a `for (const char of propDefString) {`, do you think that would be a good change to make?

ysugimoto replied:
With thinking of loop performance, I chose primitive for-loop but I didn't realize ES6 syntax for-of loop. it looks neat! I'll use this, thanks!

ysugimoto commented:
I changed the approach to deal with parsing multi-line definition:

## Not to do in coercing phase, do in unified transforming phase

I added more tree-mutator, `setNestedMultilineProperties`, which analyze AST and transform to expected property nodes for nested property definition.
To being transformed by `setNestedMultilineProperties`, it has the additional guarantee that all nested property can be dealt with. Therefore, the `property-coercer`can keep it simple -- especially String coercer doesn't need to consider `hasMany` or not anymore. 

In addition, I included the fix of https://trello.com/c/njf30klB/220-error-if-description-is-in-multiple-blocks. If the user wrote descriptions in multiple blocks, then we'll raise an error

ysugimoto commented on packages/tc-markdown-parser/__tests__/first-paragraph.spec.js:
```diff
@@ -53,3 +53,27 @@ test('top level content in an h2-range is not parsed as description', async () =
+		description: 'this is first description',
+		someString: "how's tricks",
+	});
+});
```
This is test for https://trello.com/c/njf30klB/220-error-if-description-is-in-multiple-blocks

ysugimoto commented on packages/tc-markdown-parser/lib/property-coercers/index.js:
```diff
@@ -37,40 +37,43 @@ module.exports = {
+		return {
+			valid: true,
+			value: flattenNodeToPlainString(subdocument),
+		};
```
No longer we don't need to consider `hasMany` in string coercer because nested definition will be processed at NestedString coercer

ysugimoto commented on packages/tc-markdown-parser/lib/tree-mutators/create-treecreeper-description-node.js:
```diff
@@ -16,6 +17,13 @@ module.exports = function createBizopsDescriptionNode({
+				message:
+					'only one description is allowed to define in top level',
+			});
+		}
```
validates desciprion node exists multiply

ysugimoto commented on packages/tc-markdown-parser/lib/tree-mutators/set-nested-multiline-properties.js:
```diff
@@ -0,0 +1,84 @@
+	return function transform(tree) {
+		visit(tree, 'property', mutate);
+	};
+};
```
All nested property definitions are processed in this tree-murator
